### PR TITLE
assignment 2

### PR DESCRIPTION
**Pull request by martinmessina13**
Integration with DummyJSON API for Contact Management

### Summary

This PR introduces full integration between Salesforce Contacts and the external DummyJSON API, automating the process of fetching and sending user information based on trigger events. A trigger handler class was added to ensure logicless and recursion-safe triggers, following best practices for maintainability, safety, and reusability.

---

### Changes

- **Introduced `DummyJSONCallout` Apex class** with robust HTTP GET and POST callout logic:
  - Supports fetching user data from DummyJSON and updating the corresponding Contact in Salesforce.
  - Supports sending Salesforce Contact data to DummyJSON for contact creation and updating.
  - Uses best practices such as exception handling, leveraging named credentials, and input validation.
  - Provides internal helper methods to map JSON data to Salesforce objects and vice versa.
  - Bulk-safe and trigger-compatible via `@future` methods for asynchronous callout support.

- **Implemented thorough test coverage**:
  - Added `DummyJSONCalloutTest` to verify all callout paths using mock responses.
  - Tests all main pathways, including callout flows, helper methods, and negative/edge cases.
  - Ensures correctness and provides guardrails for future changes.

- **Created a new trigger handler class (`ContactTriggerHandler`)** for logicless and recursion-safe trigger architecture:
  - Decouples business logic from the `ContactTrigger`, improving readability, maintainability, and testability.
  - Implements methods for:
    - Generating `DummyJSON_Id__c` on insert.
    - Handling callout execution to `DummyJSONCallout` based on Contact record state (insert/update and value of `DummyJSON_Id__c`).
  - **Explicitly designed to prevent trigger recursion** by isolating, controlling, and deduplicating logic execution in accordance with Salesforce best practices.

- **Updated `ContactTrigger`** to delegate all processing to the handler:
  - Sets the random `DummyJSON_Id__c` for new Contacts when not provided.
  - Initiates the correct `DummyJSONCallout` action based on Contact record state and value of `DummyJSON_Id__c`.

- **General improvements:**
  - Enforced bulk-safe and governor-limit-safe logic throughout.
  - Improved field-level security checks.
  - Provided detailed documentation within code, following Salesforce integration best practices.

---

### New Handler Class Details

- **ContactTriggerHandler:**
  - Centralizes all trigger logic, making the trigger itself logicless and recursion-safe.
  - Implements methods for required business logic while protecting against recursive execution and enforcing best practices for trigger design.
  - Simplifies future testability and maintainability.

---

### Testing

- Added, expanded, and refactored tests for:
  - Callout logic (GET/POST and error cases)
  - JSON serialization/deserialization helpers
  - Trigger behaviors (insert, update, and edge cases)
  - Handler functionality and recursion prevention logic
- Achieves comprehensive code coverage and regression safety.

## Summary by CodeRabbit

* **New Features**
  * Introduced automated integration between Contact records and an external DummyJSON API, including batch processing for user data retrieval and creation.
  * Added a custom logging object with fields for log level, message, and stack trace to capture and store application errors.
  * Expanded Contact trigger to handle before/after insert and before update events for enhanced automation.
  * Added authentication configuration and improved file visibility settings for development.
  * Enhanced HTTP callout mocking to support configurable success, error responses, and simulated connection failures for more robust testing.

* **Bug Fixes**
  * Improved error handling and permission checks during API callouts and data processing.

* **Tests**
  * Updated test methods to support new bulk processing and method signatures.
  * Added new tests covering error scenarios and exception handling for API callouts.

### CODE REVIEW

File: force-app/main/default/classes/DummyJSONCallout.cls [hunk 1]
```diff
@@ -124,13 +208,51 @@
...
+			'lastName' => contact.LastName != null ? contact.LastName : 'Unknown', 
+			'email' => contact.Email != null ? contact.Email : 'Unknown', 
+			'phone' => contact.Phone != null ? contact.Phone : 'Unknown'
+		};
+		if(String.isBlank(contact.Email)){
+			throw new DummyJSONException('Contact does not have an associated email address');
+		}
+		if(String.isBlank(contact.Phone)){
+			throw new DummyJSONException('Contact does not have a phone number');
+		}
```
[the remark this answers is not in the document]
Comment (martinmessina13) [reply force-app/main/default/classes/DummyJSONCallout.cls]: I added the checks but the exercise requests to default the values to 'Unknown' if not set.

File: force-app/main/default/classes/DummyJSONCallout.cls [hunk 2]
```diff
@@ -124,13 +208,55 @@
...
+		Contact contact = [ 
+			SELECT Id, FirstName, LastName, Email, Phone 
+			FROM Contact WHERE 
+			Id = : contactId
+			WITH USER_MODE
+		];
+
+		if(contact == null){
+			throw new DummyJSONException('Contact not found with ID: ' + contactId);
+		}
```
[the remark this answers is not in the document]
Comment (martinmessina13) [reply force-app/main/default/classes/DummyJSONCallout.cls]: Done.

**Comment by martinmessina13** [comment]
I’m not sure why the apex tests job is failing in this PR. All the individual tests are passing with 100% pass rate, and org-wide code coverage is above the required threshold. If anyone has insights into why the jobs are still being marked as failed, I’d appreciate your help!
